### PR TITLE
Multiple function shared lib

### PR DESCRIPTION
**Pull request by gbaraldi**
Option to compile a shared lib with multiple functions. The API receives a vector of tuples that look like `Tuple{func,Tuple{args}`. Ideally this would receive a module and compile the methods defined in the module, but I still have to figure out how to do that.

### CODE REVIEW

**Review by brenhinkeller — COMMENTED** [review]
Want to add a simple test for these new methods?

**Comment by chriselrod** [comment]
> Ideally this would receive a module and compile the methods defined in the module, but I still have to figure out how to do that.

All methods in a module? How would you determine function-argument pairs, and name mangling if there are multiple arguments for the same function?

**Comment by gbaraldi** [comment]
@chriselrod The perfect API in my mind would be something like
```julia
module SquareLib

@noinline square(n::Float64) = n*n

function squaresquare(n::Float64)
    square(square(n))
end

function squaresquaresquare(n::Float64)
    square(squaresquare(n))
end

end

compile_shlib(SquareLib)
```
So all methods have to have concrete types as arguments, have different names etc. The issue is I don't know how I would do that. Maybe instead of a module I would use a macro and a scope or something.

**Comment by chriselrod** [comment]
C++ has `extern "C" { }` for this.

What about just looking for all `Base.@ccallable` functions?

**Comment by brenhinkeller** [comment]
Hmm, looks like there's maybe an issue with the name (de?)mangling in the test? You can ignore "integration nightly" though

**Comment by gbaraldi** [comment]
Looking for the ccallable functions is an option, I'm not sure how I would do that since I think it only changes a bit in the `MethodInstance` but I only need the `MethodInstance` anyway

File: src/StaticCompiler.jl [hunk 1]
```diff
@@ -460,9 +459,83 @@ function native_code_native(@nospecialize(f), @nospecialize(tt), name = GPUCompi
...
+    f,tt = funcs[1]
+    mod = native_llvm_module(f,tt, kwargs...)
+    if length(funcs) > 1
+        for func in funcs[2:end]
+            @show f,tt = func
+            tmod = native_llvm_module(f,tt, kwargs...)
+            link!(mod,tmod)
+        end
+    end
+    if mangle_names
```
Comment (brenhinkeller) [inline src/StaticCompiler.jl:473]: The [terminology](https://en.wikipedia.org/wiki/Name_mangling) is a bit new to me, but if I'm reading that right, I think this is technically demangling?

File: src/StaticCompiler.jl [hunk 2]
```diff
@@ -460,9 +459,83 @@ function native_code_native(@nospecialize(f), @nospecialize(tt), name = GPUCompi
...
+        for func in functions(mod)
+            fname = name(func)
+            if fname[1:6] == "julia_"
+                name!(func,fname[7:end])
+            end
+        end
+    end
```
Comment (brenhinkeller) [inline src/StaticCompiler.jl:480]: The [terminology](https://en.wikipedia.org/wiki/Name_mangling) is a bit new to me, but if I'm reading that right, I think this is technically demangling?

**Comment by brenhinkeller** [comment]
Tests look good now!

File: src/StaticCompiler.jl (hunk 2 above)
Comment (gbaraldi) [reply src/StaticCompiler.jl:480]: you might be right actually 😄 . I will change it to `demangle`

**Comment by MasonProtter** [comment]
Yeah, this looks good @gbaraldi. Do you have any interest in trying to make the `compile` interface also work with this? No problem if not

**Comment by gbaraldi** [comment]
@MasonProtter I will take a look tomorrow, since you call into `generate_obj` it's just a question of adding a function and some way to serialize multiple compiled functions.

File: test/testcore.jl [hunk 3]
```diff
@@ -292,3 +292,29 @@ end
...
+
+function squaresquaresquare(n)
+    square(squaresquare(n))
+end
+
+@testset "Multiple Function Dylibs" begin
+
+
+    funcs = [(squaresquare,(Float64,)), (squaresquaresquare,(Float64,))]
+    filepath = compile_shlib(funcs, mangle_names=true)
```
Comment (brenhinkeller) [inline test/testcore.jl:310]: Looks like this just needs to change to `demangle` too

Comment (gbaraldi) [reply test/testcore.jl:310]: As it turns out, find and replace is hard :)

**Comment by brenhinkeller** [comment]
Still thinking about the `compile` interface, or want to merge this without and think about that later?

**Comment by brenhinkeller** [comment]
I'm about to disappear for fieldwork, so no rush on my end but I probably won't be able to be very responsive for a bit after tomorrow

**Comment by MasonProtter** [comment]
Okay, let’s just merge this and deal with `compile` in a later PR

**Pull request merged**